pydm/widgets/rules.py

```diff
@@ -0,0 +1,155 @@
+            c = PyDMChannel(ch['channel'],
+                            connection_slot=partial_connection,
+                            value_slot=partial_value)
+            if is_pydm_app():
+                self.app.add_connection(c)
+            self.channels.append(c)
+
+    def channel_conn_callback(self, index, channel_name, trigger, value):
+        """
+        Callback executed when a channel connection status is changed.
+
+        Parameters
+        ----------
+        index : int
+            The channel index on the list for this rule.
+        channel_name : str
+            The channel address.
+        trigger : bool
+            Whether or not this channel should trigger a calculation of the
+            expression
+        value : bool
+            Whether or not this channel is connected.
+
+        Returns
+        -------
+        None
+        """
+        self.channels_connection[index] = value
+
+    def channel_value_callback(self, index, channel_name, trigger, value):
+        """
+        Callback executed when a channel receives a new value.
+
+        Parameters
+        ----------
+        index : int
+            The channel index on the list for this rule.
+        channel_name : str
+            The channel address.
+        trigger : bool
+            Whether or not this channel should trigger a calculation of the
+            expression
+        value : any
+            The new value for this channel.
+
+        Returns
+        -------
+        None
+        """
+        self.channels_value[index] = value
+        if trigger:
+            if not all(self.channels_connection):
+                logger.error(
+                    "Rule %s: Not all channels are connected, skipping execution.",
+                    self.name)
+                return
+            self.should_calculate = True
+
+    def run(self):
+        """
+        Main loop of the RulesEngine which runs at 30Hz until a interruption is
+        requested and calculates the expression if a new value is available at
+        one of the trigger channels.
+
+        Returns
+        -------
+        None
+        """
+        while not self.isInterruptionRequested():
+            if self.should_calculate:
+                self.calculate_expression()
+            QThread.msleep(33)  # 30Hz
+
+    def calculate_expression(self):
+        """
+        Evaluate the expression defined by the rule and emit the `rule_signal`
+        with the new value.
+
+        Returns
+        -------
+        None
+        """
+        self.should_calculate = False
+        ch = self.channels_value
+        try:
+            val = eval(self.expression)
+            payload = {'name': self.name, 'property': self.property, 'value': val}
+            self.rule_signal.emit(payload)
+        except Exception as e:
+            logger.error("Error while evaluating Rule. Exception was: %s", e)
```

Comment:
Might want to swap this to `logger.exception` as well. User will want to know what went wrong with their rule.

Reply:
Will do! 👍